integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java

```diff
@@ -203,9 +203,67 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
         }
     }
 
+    @Test
+    public void offsetManuallyManaged(KafkaCluster cluster, Admin admin) throws Exception {
+        admin.createTopics(List.of(new NewTopic(TOPIC_1, 1, (short) 1))).all().get();
+
+        String config = """
+                proxy:
+                  address: %s
+                clusters:
+                  demo:
+                    bootstrap_servers: %s
+                filters:
+                - type: ApiVersions
+                - type: BrokerAddress
+                """.formatted(PROXY_ADDRESS, cluster.getBootstrapServers());
+
+        try (var proxy = startProxy(config)) {
+            try (var producer = new KafkaProducer<String, String>(Map.of(
+                    ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, PROXY_ADDRESS,
+                    ProducerConfig.CLIENT_ID_CONFIG, "offsetManuallyManaged",
+                    ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
+                    ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
+                    ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG, 3_600_000))) {
+                producer.send(new ProducerRecord<>(TOPIC_1, "my-key", MESSAGE_TEXT)).get();
+                producer.send(new ProducerRecord<>(TOPIC_1, "my-key2", MESSAGE_TEXT)).get();
+                producer.flush();
+            }
+
+            ConsumerRecords<String, String> records1;
+            long offsetCommitted;
+            long actualEndOffset;
+            long offsetToCommit;
+            try (var consumer = new KafkaConsumer<String, String>(Map.of(
+                    ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, PROXY_ADDRESS,
+                    ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class,
+                    ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class,
+                    ConsumerConfig.GROUP_ID_CONFIG, "my-group-id",
+                    ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false",
+                    ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"))) {
+
+                Set<TopicPartition> partitions = new HashSet<>();
+                TopicPartition actualTopicPartition = new TopicPartition(TOPIC_1, 0);
+                partitions.add(actualTopicPartition);
+                consumer.assign(Set.of(actualTopicPartition));
+
+                records1 = consumer.poll(Duration.ofSeconds(100));
+
+                actualEndOffset = consumer.endOffsets(partitions).get(actualTopicPartition);
+                offsetToCommit = actualEndOffset - 1;
+                consumer.commitSync(Map.of(actualTopicPartition, new OffsetAndMetadata(offsetToCommit)));
+
+                var committed = consumer.committed(partitions);
+                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
+                offsetCommitted = offsetAndMetadata.offset();
+            }
+            assertEquals(offsetCommitted, offsetToCommit);
```

Comment:
Yes, this is pretty much what I expected. I'd refactor it a little bit.
1. We don't need to poll for records, actually. All we need is to commit an offset to broker
2. Verify the offsetCommitted is changed after we committed (i.e., make sure kproxy indeed pass the correct info to the broker)

```java
                // before committing the offset, make sure the offset stored in broker is 0 (or -1 or exception (maybe assertThrow to verify?))
                var beforeCommitted = consumer.committed(partitions);
                assertEquals(0, beforeCommitted);

                // hard code the commit to offset
                offsetToCommit = 2;
                consumer.commitSync(Map.of(actualTopicPartition, new OffsetAndMetadata(offsetToCommit)));

                var committed = consumer.committed(partitions);
                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
                offsetCommitted = offsetAndMetadata.offset();
            }
            assertEquals(offsetCommitted, offsetToCommit);
```

WDYT?

Reply:
Great idea to assert before commit the offset. Thanks!